### Process of cleaning data 

#### Step 1: Data bilan tanishish

##### CSV faylni ochib, birinchi 10 qatorni ko‘rish. 

In [2]:
import pandas as pd
df_dirty=pd.read_csv(r'D:\MAAB\PYTHON\hmw\MAAB-PYTHON\project\super_dirty_students.csv')
df_dirty

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
0,1,Claudia Short,20,NaN,NaN,+1-619-379-4152x102,Katieland,someonegmail.com,1662247364,Data Science,NaN,active,3.72,good,$135,1629312830,"Apartment 37, South Kevin district, Tashkent, ...","{'hobbies': ['gun', 'nice'], 'skills': {'tech'..."
1,2,NaN,20,Female,ninety,NaN,Dawnburgh,psmith@chen.com,2017/08/29,Data Science,NaN,active,1.88,excellent,$152,11/10/2001 04:19 AM,UZ 100332 Tashkent South Patricia,"{hobbies:['against', 'good']}"
2,3,Kathryn Moyer,20,NaN,ninety,NaN,Lake Stevenmouth,NaN,2017-08-14,DATA SCIENCE,NaN,pending,NaN,excellent,185.0,1657837622,"Wendyshire 12-kv, dom 1, Tashkent","{'hobbies': ['fast', 'clearly'], 'skills': {'t..."
3,4,Ruben Wilson,twenty,fmale,81,NaN,Port Pamelafort,special,1973-09-17,data-sciens,NaN,inactive,0,good,175 USD,1682795130,"Apartment 16, North Tamara district, Tashkent,...","{'hobbies': ['left', 'role'], 'skills': {'tech..."
4,5,Robert Pruitt,20,Female,NaN,001-182-659-5631x02803,Kingburgh,six,2023/11/25,data_sciense,NaN,pending,4.37,average,$282,2013-10-13,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['sort', 'science'], 'skills': {'t..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,996,Gary Sawyer,22,fmale,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992.11.13,DATA SCIENCE,NaN,inactive,NaN,excellent,$64,14/10/2007 06:53 PM,"Apartment 12, Laurenhaven district, Tashkent, ...","{'hobbies': ['collection', 'his'], 'skills': {..."
996,997,Scott Clark,20,NaN,NaN,001-205-065-8737,Rogerborough,katiepage@hensley.com,2024.10.05,DATA SCIENCE,NaN,pending,"3,7",NaN,143,2021-06-21T07:46:27,"Patrickmouth 4-kv, dom 6, Tashkent",INVALID_JSON_DATA
997,998,NaN,20,MALE,NaN,264-577-8585,South Mary,someonegmail.com,19/11/2021 02:23 AM,data_sciense,95,inactive,-2,excellent,79.0,1975/05/11,"Apartment 38, New Brendan district, Tashkent, ...","{'hobbies': ['street', 'read'], 'skills': {'te..."
998,999,Brittany Barrett,21,FEMALE,86,001-663-118-1327x207,Port Rachael,richardgreen@@shannon-jenkins.info,1625723463,data_sciense,110%,active,3.06,average,"238,00",2015.05.24,"25998 Martinez Grove Apt. 473, West Scott","{'hobbies': ['game', 'total'], 'skills': {'tec..."


In [3]:
df_dirty.head(10)

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
0,1,Claudia Short,20,NaN,NaN,+1-619-379-4152x102,Katieland,someonegmail.com,1662247364,Data Science,NaN,active,3.72,good,$135,1629312830,"Apartment 37, South Kevin district, Tashkent, ...","{'hobbies': ['gun', 'nice'], 'skills': {'tech'..."
1,2,NaN,20,Female,ninety,NaN,Dawnburgh,psmith@chen.com,2017/08/29,Data Science,NaN,active,1.88,excellent,$152,11/10/2001 04:19 AM,UZ 100332 Tashkent South Patricia,"{hobbies:['against', 'good']}"
2,3,Kathryn Moyer,20,NaN,ninety,NaN,Lake Stevenmouth,NaN,2017-08-14,DATA SCIENCE,NaN,pending,NaN,excellent,185.0,1657837622,"Wendyshire 12-kv, dom 1, Tashkent","{'hobbies': ['fast', 'clearly'], 'skills': {'t..."
3,4,Ruben Wilson,twenty,fmale,81,NaN,Port Pamelafort,special,1973-09-17,data-sciens,NaN,inactive,0,good,175 USD,1682795130,"Apartment 16, North Tamara district, Tashkent,...","{'hobbies': ['left', 'role'], 'skills': {'tech..."
4,5,Robert Pruitt,20,Female,NaN,001-182-659-5631x02803,Kingburgh,six,2023/11/25,data_sciense,NaN,pending,4.37,average,$282,2013-10-13,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['sort', 'science'], 'skills': {'t..."
5,6,David Martinez,20,Male,ninety,001-074-828-6016x937,Hugheshaven,nathangibbs@hotmail.com,1603493647,DATA SCIENCE,110,pending,15.5,average,"85,00",1983-04-30T01:46:36,"Apartment 42, North Jamesfurt district, Tashke...","{hobbies:['order', 'keep']}"
6,7,NaN,18 years,FEMALE,71,314.018.0928x3134,Sharpfurt,nguyenlisa@@horn.org,1982-01-16,data science,NaN,active,0,excellent,275.0,19/11/2004 11:42 PM,"Lake Tracishire 29-kv, dom 8, Tashkent","{'hobbies': ['fall', 'teacher'], 'skills': {'t..."
7,8,Nicholas Dennis,21,FEMALE,ninety,8091556011,Matthewfort,avoid,2000.06.26,python,63,active,four point five,excellent,135,1978/09/23,"70576 Butler Harbor Suite 847, Janetberg",INVALID_JSON_DATA
8,9,NaN,twenty,fmale,77,854-964-5423,South Paulchester,stacysmith@taylor.biz,2000-03-15,ds,110%,active,15.5,excellent,104 USD,1995-10-20,"Williamsborough 2-kv, dom 18, Tashkent","{'hobbies': ['require', 'open'], 'skills': {'t..."
9,10,Elizabeth Villegas,22,femlae,NaN,(205)467-4476x0583,South Dianatown,anthonywalker@@gmail.com,2013-02-11T07:12:29,data_sciense,99,inactive,"3,7",average,"215,00",2014-12-16T06:44:51,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['management', 'to'], 'skills': {'..."


In [4]:
df_dirty.tail(5)

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
995,996,Gary Sawyer,22,fmale,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992.11.13,DATA SCIENCE,NaN,inactive,NaN,excellent,$64,14/10/2007 06:53 PM,"Apartment 12, Laurenhaven district, Tashkent, ...","{'hobbies': ['collection', 'his'], 'skills': {..."
996,997,Scott Clark,20,NaN,NaN,001-205-065-8737,Rogerborough,katiepage@hensley.com,2024.10.05,DATA SCIENCE,NaN,pending,"3,7",NaN,143,2021-06-21T07:46:27,"Patrickmouth 4-kv, dom 6, Tashkent",INVALID_JSON_DATA
997,998,NaN,20,MALE,NaN,264-577-8585,South Mary,someonegmail.com,19/11/2021 02:23 AM,data_sciense,95,inactive,-2,excellent,79.0,1975/05/11,"Apartment 38, New Brendan district, Tashkent, ...","{'hobbies': ['street', 'read'], 'skills': {'te..."
998,999,Brittany Barrett,21,FEMALE,86,001-663-118-1327x207,Port Rachael,richardgreen@@shannon-jenkins.info,1625723463,data_sciense,110%,active,3.06,average,"238,00",2015.05.24,"25998 Martinez Grove Apt. 473, West Scott","{'hobbies': ['game', 'total'], 'skills': {'tec..."
999,1000,NaN,twenty,NaN,ninety,486.754.8735,Dariusborough,?ali@@gmail..com,1624533231,PYTHNO,70,inactive,four point five,excellent,224.0,27/06/2015 08:04 AM,"Apartment 24, Grossbury district, Tashkent, UZ...",INVALID_JSON_DATA


In [5]:
df_dirty.sample(3)

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
148,149,NaN,18 years,male,NaN,266.209.7518,Shannonhaven,?ali@@gmail..com,2005.08.23,ds,110%,active,four point five,NaN,$297,1980.01.05,"Ewingview 29-kv, dom 16, Tashkent","{'hobbies': ['possible', 'should'], 'skills': ..."
991,992,NaN,21,FEMALE,73,545-622-4455,East Rebeccashire,TV,1990.07.13,DATA SCIENCE,75,active,NaN,excellent,$140,2008.10.22,"1137 Victor Garden, Crosbymouth","{hobbies:['level', 'owner']}"
423,424,Christopher Gill,22,FEMALE,NaN,001-231-307-4684x3769,Jessicaview,bassmichael@hotmail.com,1985-09-10T16:33:19,DATA SCIENCE,60,inactive,-2,excellent,218,2004/06/28,"Apartment 23, East Chad district, Tashkent, UZ...","{""hobbies"": [""world"", ""computer""], ""skills"": {..."


##### Column nomlari va sonini tekshirish. 

In [6]:
df_dirty.shape

(1000, 18)

In [7]:
print("Columns:")
print(df_dirty.columns.tolist())

Columns:
['student_id', 'name', 'age', 'gender', 'score', 'phone', 'city', 'email', 'date_of_join', 'course', 'attendance', 'status', 'gpa', 'remarks', 'money_spent', 'event_time', 'address_raw', 'profile_json']


##### Har bir columndagi missing values sonini aniqlash. 

In [8]:
df_dirty.isnull().sum()

student_id        0
name            335
age             108
gender          219
score           246
phone           371
city              0
email           151
date_of_join      0
course            0
attendance      326
status            0
gpa             141
remarks         204
money_spent       0
event_time        0
address_raw       0
profile_json      0
dtype: int64

##### Data turlari (string, numeric, date, json) haqida tushuncha hosil qilish. 

In [9]:
df_dirty.describe(include='all')

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
count,1000.000000,665,892,781,754,629,1000,849,1000,1000,674,1000,859,796,1000,1000,1000,1000
unique,NaN,665,8,8,52,629,969,567,995,10,62,3,111,4,707,999,790,815
top,NaN,Claudia Short,20,fmale,ninety,+1-619-379-4152x102,New Michael,?ali@@gmail..com,2010/06/30,data science,110%,inactive,15.5,excellent,"183,00",1983.12.01,"BROKEN,ADDRESS,DATA,,,",INVALID_JSON_DATA
freq,NaN,1,119,114,251,1,4,139,2,107,342,362,160,208,5,2,211,186
mean,500.500000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,288.819436,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,1.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,250.750000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,500.500000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,750.250000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [10]:
df_dirty.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   student_id    1000 non-null   int64
 1   name          665 non-null    str  
 2   age           892 non-null    str  
 3   gender        781 non-null    str  
 4   score         754 non-null    str  
 5   phone         629 non-null    str  
 6   city          1000 non-null   str  
 7   email         849 non-null    str  
 8   date_of_join  1000 non-null   str  
 9   course        1000 non-null   str  
 10  attendance    674 non-null    str  
 11  status        1000 non-null   str  
 12  gpa           859 non-null    str  
 13  remarks       796 non-null    str  
 14  money_spent   1000 non-null   str  
 15  event_time    1000 non-null   str  
 16  address_raw   1000 non-null   str  
 17  profile_json  1000 non-null   str  
dtypes: int64(1), str(17)
memory usage: 476.6 KB


In [11]:
df_dirty.describe()

,student_id
count,1000.000000
mean,500.500000
std,288.819436
min,1.000000
25%,250.750000
50%,500.500000
75%,750.250000
max,1000.000000


In [12]:
df_dirty.dtypes

student_id      int64
name              str
age               str
gender            str
score             str
phone             str
city              str
email             str
date_of_join      str
course            str
attendance        str
status            str
gpa               str
remarks           str
money_spent       str
event_time        str
address_raw       str
profile_json      str
dtype: object

#### Step 2: Oddiy string tozalash 

##### Har bir string columnning bo‘sh joylarini olib tashlash (strip). 

In [13]:
string_col = df_dirty.select_dtypes(include='object').columns
string_col

C:\Users\user\AppData\Local\Temp\ipykernel_10860\1853031010.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  string_col = df_dirty.select_dtypes(include='object').columns


Index(['name', 'age', 'gender', 'score', 'phone', 'city', 'email',
       'date_of_join', 'course', 'attendance', 'status', 'gpa', 'remarks',
       'money_spent', 'event_time', 'address_raw', 'profile_json'],
      dtype='str')

In [14]:
for col in string_col:
    df_dirty[col] = df_dirty[col].str.strip()

##### Bo‘sh stringlarni NaN yoki Null bilan almashtirish. 

In [15]:
import numpy as np

correction age

In [16]:
df_dirty['age']=df_dirty['age'].replace('twenty', 20)

In [17]:
df_dirty['age']=df_dirty['age'].replace('18 years', 18)

correction gender

In [18]:
df_dirty['gender']=df_dirty['gender'].str.lower()

In [19]:
df_dirty['gender']=df_dirty['gender'].replace('fmale', 'female')

In [20]:
df_dirty['gender']=df_dirty['gender'].replace('fmale', 'female')

In [21]:
df_dirty['gender']=df_dirty['gender'].replace('femlae', 'female')

correction phone


In [22]:
df_dirty['phone'] = df_dirty['phone'].str.replace(r'[.*x]', '-', regex=True)

correction score

In [23]:
df_dirty['score']=df_dirty['score'].replace('ninety', 90)

correction course

In [24]:
df_dirty['course']=df_dirty['course'].str.lower()

In [25]:
df_dirty['course']=df_dirty['course'].replace('d.s.', 'data science')

In [26]:
df_dirty['course']=df_dirty['course'].replace('data_sciense', 'data science')

In [27]:
df_dirty['course']=df_dirty['course'].replace('data-sciens', 'data science')

In [28]:
df_dirty['course']=df_dirty['course'].replace('pyhton', 'python')

In [29]:
df_dirty['course']=df_dirty['course'].replace('pythno', 'python')

correction attendance

In [30]:
df_dirty['attendance'] = df_dirty['attendance'].astype(str).str.strip().str.replace('%', '')

gpa correction

In [31]:
df_dirty['gpa']=df_dirty['gpa'].replace('four point five', 4.5)

In [32]:
df_dirty['gpa']=df_dirty['gpa'].replace('3,7', 3.7)

In [33]:
df_dirty['gpa']=df_dirty['gpa'].replace('15,5', 1.55)

In [34]:
df_dirty['gpa']=df_dirty['gpa'].replace('-2', 2)

correction money

In [35]:
import re

def clean_money(value):
    if pd.isna(value):
        return None
    
    value = str(value).strip()
    
    value = re.sub(r'[^\d,.]', ' ', value.replace('$', '').replace('USD', ''))
    
    value = value.strip()
    
    value = value.replace(',', '.')
    
    try:
        return float(value)
    except:
        return None

df_dirty['money_spent'] = df_dirty['money_spent'].apply(clean_money)

In [36]:
#df_dirty['age'].isna().sum()

In [37]:
#df_dirty['course'].isna().sum()

In [38]:
#df_dirty['gender'].isna().sum()

#### Step 3: Raqam va sanalarni tozalash 

##### Age, score, GPA, attendance, money_spent kabi numeric columnlarni tozalash va int yoki float formatiga keltirish. 

In [39]:
df_dirty.dtypes

student_id        int64
name                str
age              object
gender              str
score            object
phone               str
city                str
email               str
date_of_join        str
course              str
attendance          str
status              str
gpa              object
remarks             str
money_spent     float64
event_time          str
address_raw         str
profile_json        str
dtype: object

In [40]:
num_cols = ['age', 'score', 'gpa', 'attendance', 'money_spent']
for col in num_cols:
    df_dirty[col] = (df_dirty[col].astype(str).str.replace(r'[^0-9.]', '', regex=True)  # raqam va nuqtadan boshqa hammasini o‘chirish
    )

In [41]:
for col in num_cols:
    df_dirty[col] = pd.to_numeric(df_dirty[col], errors='coerce')

In [42]:
df_dirty.dtypes

student_id        int64
name                str
age             float64
gender              str
score           float64
phone               str
city                str
email               str
date_of_join        str
course              str
attendance      float64
status              str
gpa             float64
remarks             str
money_spent     float64
event_time          str
address_raw         str
profile_json        str
dtype: object

Logic here is important too

In [43]:
df_dirty.loc[(df_dirty['age'] < 0) | (df_dirty['age'] > 120), 'age'] = np.nan # inson yoshi 120 dan yosh va 0 dan katta bo'lishi kerak

In [44]:
df_dirty.loc[(df_dirty['gpa'] < 0) | (df_dirty['gpa'] > 4.5), 'gpa'] = np.nan # GPA max 4.5 va min 0 bo'lishi mumkin


In [45]:
df_dirty.loc[(df_dirty['score'] < 0) | (df_dirty['score'] > 100), 'score'] = np.nan # Score 0 dan kichkina va 100 dan katta bo'lishi mumkin emas

In [46]:
df_dirty.loc[(df_dirty['attendance'] < 0 | (df_dirty['attendance'] > 100)), 'attendance'] = np.nan # attandance max noma'lum lekin min 0 bo'lishi kerak

attendanceni mean bilan to'ldirish

In [47]:
df_dirty['attendance'] = df_dirty['attendance'].fillna(df_dirty['attendance'].mean()).astype(int)

In [48]:
df_dirty[num_cols].isnull().sum()

age            108
score          246
gpa            301
attendance       0
money_spent      0
dtype: int64

In [49]:
df_dirty

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
0,1,Claudia Short,20.0,NaN,NaN,+1-619-379-4152-102,Katieland,someonegmail.com,1662247364,data science,100,active,3.72,good,135.0,1629312830,"Apartment 37, South Kevin district, Tashkent, ...","{'hobbies': ['gun', 'nice'], 'skills': {'tech'..."
1,2,NaN,20.0,female,90.0,NaN,Dawnburgh,psmith@chen.com,2017/08/29,data science,100,active,1.88,excellent,152.0,11/10/2001 04:19 AM,UZ 100332 Tashkent South Patricia,"{hobbies:['against', 'good']}"
2,3,Kathryn Moyer,20.0,NaN,90.0,NaN,Lake Stevenmouth,NaN,2017-08-14,data science,100,pending,NaN,excellent,185.0,1657837622,"Wendyshire 12-kv, dom 1, Tashkent","{'hobbies': ['fast', 'clearly'], 'skills': {'t..."
3,4,Ruben Wilson,20.0,female,81.0,NaN,Port Pamelafort,special,1973-09-17,data science,100,inactive,0.00,good,175.0,1682795130,"Apartment 16, North Tamara district, Tashkent,...","{'hobbies': ['left', 'role'], 'skills': {'tech..."
4,5,Robert Pruitt,20.0,female,NaN,001-182-659-5631-02803,Kingburgh,six,2023/11/25,data science,100,pending,4.37,average,282.0,2013-10-13,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['sort', 'science'], 'skills': {'t..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,996,Gary Sawyer,22.0,female,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992.11.13,data science,100,inactive,NaN,excellent,64.0,14/10/2007 06:53 PM,"Apartment 12, Laurenhaven district, Tashkent, ...","{'hobbies': ['collection', 'his'], 'skills': {..."
996,997,Scott Clark,20.0,NaN,NaN,001-205-065-8737,Rogerborough,katiepage@hensley.com,2024.10.05,data science,100,pending,3.70,NaN,143.0,2021-06-21T07:46:27,"Patrickmouth 4-kv, dom 6, Tashkent",INVALID_JSON_DATA
997,998,NaN,20.0,male,NaN,264-577-8585,South Mary,someonegmail.com,19/11/2021 02:23 AM,data science,95,inactive,2.00,excellent,79.0,1975/05/11,"Apartment 38, New Brendan district, Tashkent, ...","{'hobbies': ['street', 'read'], 'skills': {'te..."
998,999,Brittany Barrett,21.0,female,86.0,001-663-118-1327-207,Port Rachael,richardgreen@@shannon-jenkins.info,1625723463,data science,110,active,3.06,average,238.0,2015.05.24,"25998 Martinez Grove Apt. 473, West Scott","{'hobbies': ['game', 'total'], 'skills': {'tec..."


##### Date columnlarni pandas datetime formatiga o‘tkazish. 

In [50]:
df_dirty

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
0,1,Claudia Short,20.0,NaN,NaN,+1-619-379-4152-102,Katieland,someonegmail.com,1662247364,data science,100,active,3.72,good,135.0,1629312830,"Apartment 37, South Kevin district, Tashkent, ...","{'hobbies': ['gun', 'nice'], 'skills': {'tech'..."
1,2,NaN,20.0,female,90.0,NaN,Dawnburgh,psmith@chen.com,2017/08/29,data science,100,active,1.88,excellent,152.0,11/10/2001 04:19 AM,UZ 100332 Tashkent South Patricia,"{hobbies:['against', 'good']}"
2,3,Kathryn Moyer,20.0,NaN,90.0,NaN,Lake Stevenmouth,NaN,2017-08-14,data science,100,pending,NaN,excellent,185.0,1657837622,"Wendyshire 12-kv, dom 1, Tashkent","{'hobbies': ['fast', 'clearly'], 'skills': {'t..."
3,4,Ruben Wilson,20.0,female,81.0,NaN,Port Pamelafort,special,1973-09-17,data science,100,inactive,0.00,good,175.0,1682795130,"Apartment 16, North Tamara district, Tashkent,...","{'hobbies': ['left', 'role'], 'skills': {'tech..."
4,5,Robert Pruitt,20.0,female,NaN,001-182-659-5631-02803,Kingburgh,six,2023/11/25,data science,100,pending,4.37,average,282.0,2013-10-13,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['sort', 'science'], 'skills': {'t..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,996,Gary Sawyer,22.0,female,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992.11.13,data science,100,inactive,NaN,excellent,64.0,14/10/2007 06:53 PM,"Apartment 12, Laurenhaven district, Tashkent, ...","{'hobbies': ['collection', 'his'], 'skills': {..."
996,997,Scott Clark,20.0,NaN,NaN,001-205-065-8737,Rogerborough,katiepage@hensley.com,2024.10.05,data science,100,pending,3.70,NaN,143.0,2021-06-21T07:46:27,"Patrickmouth 4-kv, dom 6, Tashkent",INVALID_JSON_DATA
997,998,NaN,20.0,male,NaN,264-577-8585,South Mary,someonegmail.com,19/11/2021 02:23 AM,data science,95,inactive,2.00,excellent,79.0,1975/05/11,"Apartment 38, New Brendan district, Tashkent, ...","{'hobbies': ['street', 'read'], 'skills': {'te..."
998,999,Brittany Barrett,21.0,female,86.0,001-663-118-1327-207,Port Rachael,richardgreen@@shannon-jenkins.info,1625723463,data science,110,active,3.06,average,238.0,2015.05.24,"25998 Martinez Grove Apt. 473, West Scott","{'hobbies': ['game', 'total'], 'skills': {'tec..."


In [51]:
df_dirty.columns

Index(['student_id', 'name', 'age', 'gender', 'score', 'phone', 'city',
       'email', 'date_of_join', 'course', 'attendance', 'status', 'gpa',
       'remarks', 'money_spent', 'event_time', 'address_raw', 'profile_json'],
      dtype='str')

In [52]:
#df_gpa=df_dirty['gpa'].groupby(by='city').mean()


In [53]:
df_dirty.columns

Index(['student_id', 'name', 'age', 'gender', 'score', 'phone', 'city',
       'email', 'date_of_join', 'course', 'attendance', 'status', 'gpa',
       'remarks', 'money_spent', 'event_time', 'address_raw', 'profile_json'],
      dtype='str')

In [54]:
#df_dirty.loc[df_dirty['date_of_join'].isna(), 'date_of_join'].head(10)
#df_dirty[['date_of_join', 'event_time']]

In [55]:
#df_dirty['date_of_join'] = pd.to_datetime(df_dirty['date_of_join'], format="%Y-%m-%d", errors='coerce')
#df_dirty['event_time'] = pd.to_datetime(df_dirty['event_time'], format="%Y-%m-%d", errors='coerce')

In [56]:
#df_dirty[['date_of_join', 'event_time']].dtypes

In [57]:
#df_dirty[['date_of_join', 'event_time']].isna().sum()

In [58]:

def preprocess_unix(val):
    try:
        s = str(val).strip()
        if s.isdigit() and len(s) == 10:
            return pd.Timestamp(int(s), unit='s')
    except:
        pass
    return val

def parse_mixed_dates(date_series):
    formats = [
        "%m/%d/%Y %I:%M %p",   # 11/10/2001 4:19 PM
        "%d/%m/%Y %I:%M %p",   # 16/07/2000 01:07 PM
        "%Y-%m-%dT%H:%M:%S",   # 2014-12-16T06:44:51
        "%Y-%m-%d %H:%M:%S",   # 2001-06-06 05:29:47
        "%m/%d/%Y",            # 9/23/1978
        "%d/%m/%Y",            # 23/09/1978
        "%Y.%m.%d",            # 2013.05.14 / 2020.11.26
        "%m/%d/%Y %H:%M",      # 5/27/2003 14:00
        "%d/%m/%Y %H:%M",      # 27/05/2003 14:00
    ]

    result = pd.Series([pd.NaT] * len(date_series), index=date_series.index)
    remaining_mask = pd.Series([True] * len(date_series), index=date_series.index)

    for fmt in formats:
        if not remaining_mask.any():
            break
        parsed = pd.to_datetime(
            date_series.where(remaining_mask),
            format=fmt,
            errors="coerce"
        )
        newly_parsed = parsed.notna() & remaining_mask
        result[newly_parsed] = parsed[newly_parsed]
        remaining_mask[newly_parsed] = False

    # Final fallback for anything remaining
    if remaining_mask.any():
        fallback = pd.to_datetime(date_series.where(remaining_mask), errors="coerce")
        result[remaining_mask] = fallback[remaining_mask]

    return result



# Step 1: Convert Unix timestamps first
df_dirty["event_time"] = df_dirty["event_time"].apply(preprocess_unix)
df_dirty["date_of_join"] = df_dirty["date_of_join"].apply(preprocess_unix)
# Step 2: Parse remaining string dates
mask = df_dirty["event_time"].apply(lambda x: not isinstance(x, pd.Timestamp))
df_dirty.loc[mask, "event_time"] = parse_mixed_dates(df_dirty.loc[mask, "event_time"])


mask = df_dirty["date_of_join"].apply(lambda x: not isinstance(x, pd.Timestamp))
df_dirty.loc[mask, "date_of_join"] = parse_mixed_dates(df_dirty.loc[mask, "date_of_join"])
# Step 3: Check results
print(df_dirty["event_time"].isna().sum(), df_dirty["date_of_join"].isna().sum() )


173 168


In [59]:
df_dirty.isna().sum()

student_id        0
name            335
age             108
gender          219
score           246
phone           371
city              0
email           151
date_of_join    168
course            0
attendance        0
status            0
gpa             301
remarks         204
money_spent       0
event_time      173
address_raw       0
profile_json      0
dtype: int64

In [60]:
df_dirty

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
0,1,Claudia Short,20.0,NaN,NaN,+1-619-379-4152-102,Katieland,someonegmail.com,2022-09-03 23:22:44,data science,100,active,3.72,good,135.0,2021-08-18 18:53:50,"Apartment 37, South Kevin district, Tashkent, ...","{'hobbies': ['gun', 'nice'], 'skills': {'tech'..."
1,2,NaN,20.0,female,90.0,NaN,Dawnburgh,psmith@chen.com,2017-08-29 00:00:00,data science,100,active,1.88,excellent,152.0,2001-11-10 04:19:00,UZ 100332 Tashkent South Patricia,"{hobbies:['against', 'good']}"
2,3,Kathryn Moyer,20.0,NaN,90.0,NaN,Lake Stevenmouth,NaN,NaT,data science,100,pending,NaN,excellent,185.0,2022-07-14 22:27:02,"Wendyshire 12-kv, dom 1, Tashkent","{'hobbies': ['fast', 'clearly'], 'skills': {'t..."
3,4,Ruben Wilson,20.0,female,81.0,NaN,Port Pamelafort,special,NaT,data science,100,inactive,0.00,good,175.0,2023-04-29 19:05:30,"Apartment 16, North Tamara district, Tashkent,...","{'hobbies': ['left', 'role'], 'skills': {'tech..."
4,5,Robert Pruitt,20.0,female,NaN,001-182-659-5631-02803,Kingburgh,six,2023-11-25 00:00:00,data science,100,pending,4.37,average,282.0,2013-10-13 00:00:00,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['sort', 'science'], 'skills': {'t..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,996,Gary Sawyer,22.0,female,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992-11-13 00:00:00,data science,100,inactive,NaN,excellent,64.0,2007-10-14 18:53:00,"Apartment 12, Laurenhaven district, Tashkent, ...","{'hobbies': ['collection', 'his'], 'skills': {..."
996,997,Scott Clark,20.0,NaN,NaN,001-205-065-8737,Rogerborough,katiepage@hensley.com,2024-10-05 00:00:00,data science,100,pending,3.70,NaN,143.0,2021-06-21 07:46:27,"Patrickmouth 4-kv, dom 6, Tashkent",INVALID_JSON_DATA
997,998,NaN,20.0,male,NaN,264-577-8585,South Mary,someonegmail.com,2021-11-19 02:23:00,data science,95,inactive,2.00,excellent,79.0,NaT,"Apartment 38, New Brendan district, Tashkent, ...","{'hobbies': ['street', 'read'], 'skills': {'te..."
998,999,Brittany Barrett,21.0,female,86.0,001-663-118-1327-207,Port Rachael,richardgreen@@shannon-jenkins.info,2021-07-08 05:51:03,data science,110,active,3.06,average,238.0,2015-05-24 00:00:00,"25998 Martinez Grove Apt. 473, West Scott","{'hobbies': ['game', 'total'], 'skills': {'tec..."


In [61]:
df_dirty['event_time']   = pd.to_datetime(df_dirty['event_time'])
df_dirty['date_of_join'] = pd.to_datetime(df_dirty['date_of_join'])

print(df_dirty[['event_time', 'date_of_join']].dtypes)
print(df_dirty[['event_time', 'date_of_join']].head(5))

event_time      datetime64[s]
date_of_join    datetime64[s]
dtype: object
           event_time        date_of_join
0 2021-08-18 18:53:50 2022-09-03 23:22:44
1 2001-11-10 04:19:00 2017-08-29 00:00:00
2 2022-07-14 22:27:02                 NaT
3 2023-04-29 19:05:30                 NaT
4 2013-10-13 00:00:00 2023-11-25 00:00:00


In [62]:
df_dirty[['event_time', 'date_of_join']].dtypes

event_time      datetime64[s]
date_of_join    datetime64[s]
dtype: object

#### Step 4: Email va phone validation 

##### Emaillarni kichik harflarga o‘tkazish va noto‘g‘ri formatdagi emaillarni aniqlash. 

In [63]:
def clean_email(value):
    if pd.isna(value):
        return np.nan
    
    value = str(value).strip().lower()
    
    # Fix common typos
    value = re.sub(r'@+', '@', value)   # @@ → @
    value = re.sub(r'\.+', '.', value)  # .. → .
    
    # Validate
    pattern = r'^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,7}$'
    
    return value if re.match(pattern, value) else np.nan

df_dirty['email'] = df_dirty['email'].apply(clean_email)

In [64]:
#df_dirty.drop(columns='email', inplace=True)

In [65]:
#df_dirty['email_cleaned'].isna().sum()

In [66]:
df_dirty

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,attendance,status,gpa,remarks,money_spent,event_time,address_raw,profile_json
0,1,Claudia Short,20.0,NaN,NaN,+1-619-379-4152-102,Katieland,NaN,2022-09-03 23:22:44,data science,100,active,3.72,good,135.0,2021-08-18 18:53:50,"Apartment 37, South Kevin district, Tashkent, ...","{'hobbies': ['gun', 'nice'], 'skills': {'tech'..."
1,2,NaN,20.0,female,90.0,NaN,Dawnburgh,psmith@chen.com,2017-08-29 00:00:00,data science,100,active,1.88,excellent,152.0,2001-11-10 04:19:00,UZ 100332 Tashkent South Patricia,"{hobbies:['against', 'good']}"
2,3,Kathryn Moyer,20.0,NaN,90.0,NaN,Lake Stevenmouth,NaN,NaT,data science,100,pending,NaN,excellent,185.0,2022-07-14 22:27:02,"Wendyshire 12-kv, dom 1, Tashkent","{'hobbies': ['fast', 'clearly'], 'skills': {'t..."
3,4,Ruben Wilson,20.0,female,81.0,NaN,Port Pamelafort,NaN,NaT,data science,100,inactive,0.00,good,175.0,2023-04-29 19:05:30,"Apartment 16, North Tamara district, Tashkent,...","{'hobbies': ['left', 'role'], 'skills': {'tech..."
4,5,Robert Pruitt,20.0,female,NaN,001-182-659-5631-02803,Kingburgh,NaN,2023-11-25 00:00:00,data science,100,pending,4.37,average,282.0,2013-10-13 00:00:00,"BROKEN,ADDRESS,DATA,,,","{'hobbies': ['sort', 'science'], 'skills': {'t..."
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,996,Gary Sawyer,22.0,female,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992-11-13 00:00:00,data science,100,inactive,NaN,excellent,64.0,2007-10-14 18:53:00,"Apartment 12, Laurenhaven district, Tashkent, ...","{'hobbies': ['collection', 'his'], 'skills': {..."
996,997,Scott Clark,20.0,NaN,NaN,001-205-065-8737,Rogerborough,katiepage@hensley.com,2024-10-05 00:00:00,data science,100,pending,3.70,NaN,143.0,2021-06-21 07:46:27,"Patrickmouth 4-kv, dom 6, Tashkent",INVALID_JSON_DATA
997,998,NaN,20.0,male,NaN,264-577-8585,South Mary,NaN,2021-11-19 02:23:00,data science,95,inactive,2.00,excellent,79.0,NaT,"Apartment 38, New Brendan district, Tashkent, ...","{'hobbies': ['street', 'read'], 'skills': {'te..."
998,999,Brittany Barrett,21.0,female,86.0,001-663-118-1327-207,Port Rachael,richardgreen@shannon-jenkins.info,2021-07-08 05:51:03,data science,110,active,3.06,average,238.0,2015-05-24 00:00:00,"25998 Martinez Grove Apt. 473, West Scott","{'hobbies': ['game', 'total'], 'skills': {'tec..."


##### Telefon raqamlarni tozalash va yagona standart formatga keltirish (masalan: +998...). 

In [67]:
def clean_phone(value):
    if pd.isna(value):
        return np.nan
    
    value = str(value).strip()
    
    # Drop extension (x followed by digits at the end)
    value = re.sub(r'x\d+$', '', value, flags=re.IGNORECASE)
    
    # Extract digits only
    digits = re.sub(r'\D', '', value)
    
    # 10 digits → no country code → add +1
    if len(digits) == 10:
        return f"+1-{digits[0:3]}-{digits[3:6]}-{digits[6:10]}"
    
    # 11 digits → starts with 1 (US)
    elif len(digits) == 11 and digits.startswith('1'):
        return f"+1-{digits[1:4]}-{digits[4:7]}-{digits[7:11]}"
    
    # 13 digits → starts with 001 → same as +1
    elif len(digits) == 13 and digits.startswith('001'):
        d = digits[3:]
        return f"+1-{d[0:3]}-{d[3:6]}-{d[6:10]}"
    
    else:
        return np.nan

# Apply
df_dirty['phone'] = df_dirty['phone'].apply(clean_phone)

# Verify
print(df_dirty['phone'].head(10))
print('\nValid:', df_dirty['phone'].notna().sum())
print('NaN:  ', df_dirty['phone'].isna().sum())

0                NaN
1                NaN
2                NaN
3                NaN
4                NaN
5                NaN
6                NaN
7    +1-809-155-6011
8    +1-854-964-5423
9                NaN
Name: phone, dtype: str

Valid: 261
NaN:   739


In [68]:
df_dirty['phone'].isna().sum()

np.int64(739)

#### Step 5: JSON parsing 

##### JSON columnlarini ochish (profile_json). 

In [69]:
import pandas as pd
import numpy as np
import json
import re

def fix_and_parse_json(value):
    if pd.isna(value):
        return None
    
    value = str(value).strip()          
    
    if value == 'INVALID_JSON_DATA':
        return None
    
    value = value.replace("'", '"')     # single → double quotes
    value = re.sub(r'(\{|,)\s*([a-zA-Z_]+)\s*:', r'\1"\2":', value)  
    
    try:
        return json.loads(value)
    except:
        return None

df_dirty['profile_parsed'] = df_dirty['profile_json'].apply(fix_and_parse_json)

print("Parsed successfully:", df_dirty['profile_parsed'].notna().sum())
print("Failed/NaN:         ", df_dirty['profile_parsed'].isna().sum())

Parsed successfully: 814
Failed/NaN:          186


##### JSON ichidagi ma’lumotlarni alohida columnlarga ajratish: hobbies, skills, family, devices. 

In [70]:
def extract_fields(parsed):
    if parsed is None:
        return pd.Series({
            'hobbies'             : np.nan,
            'skills_python'       : np.nan,
            'skills_excel'        : np.nan,
            'skills_sql'          : np.nan,
            'skills_soft'         : np.nan,
            'family_siblings'     : np.nan,
            'family_income_father': np.nan,
            'family_income_mother': np.nan,
            'devices_laptop_brand': np.nan,
            'devices_laptop_year' : np.nan,
            'devices_phone_brand' : np.nan,
            'devices_phone_year'  : np.nan,
        })
    
    hobbies = ', '.join(parsed.get('hobbies', [])) or np.nan
    skills  = parsed.get('skills', {})
    tech    = skills.get('tech', {})
    soft    = skills.get('soft', [])
    family  = parsed.get('family', {})
    income  = family.get('income', {})
    devices = parsed.get('devices', [])
    laptop  = next((d for d in devices if d.get('type') == 'laptop'), {})
    phone   = next((d for d in devices if d.get('type') == 'phone'),  {})

    return pd.Series({
        'hobbies'             : hobbies,
        'skills_python'       : tech.get('python', np.nan),
        'skills_excel'        : tech.get('excel',  np.nan),
        'skills_sql'          : tech.get('sql',    np.nan),
        'skills_soft'         : ', '.join(soft) if soft else np.nan,
        'family_siblings'     : family.get('siblings', np.nan),
        'family_income_father': income.get('father', np.nan),
        'family_income_mother': income.get('mother', np.nan),
        'devices_laptop_brand': laptop.get('brand', np.nan),
        'devices_laptop_year' : laptop.get('year',  np.nan),
        'devices_phone_brand' : phone.get('brand',  np.nan),
        'devices_phone_year'  : phone.get('year',   np.nan),
    })

json_expanded = df_dirty['profile_parsed'].apply(extract_fields)
print(json_expanded[['hobbies', 'skills_python', 'family_siblings', 'devices_laptop_brand']].head(10))

          hobbies  skills_python  family_siblings devices_laptop_brand
0       gun, nice            2.0              4.0                   HP
1   against, good            NaN              NaN                  NaN
2   fast, clearly            5.0              1.0                   HP
3      left, role            3.0              4.0               Lenovo
4   sort, science            1.0              3.0               Lenovo
5     order, keep            NaN              NaN                  NaN
6   fall, teacher            4.0              3.0                   HP
7             NaN            NaN              NaN                  NaN
8   require, open            1.0              0.0                 Dell
9  management, to            1.0              3.0                 Dell


##### JSON ichidagi array yoki nested objectlarni tekshirish va flatten qilish. 

In [71]:
df_dirty = pd.concat([df_dirty, json_expanded], axis=1)
df_dirty.drop(columns=['profile_json', 'profile_parsed'], inplace=True)

print("Final shape:", df_dirty.shape)
print("hobbies non-null:", df_dirty['hobbies'].notna().sum())
print("skills_python non-null:", df_dirty['skills_python'].notna().sum())

Final shape: (1000, 29)
hobbies non-null: 814
skills_python non-null: 594


In [72]:
df_dirty

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,...,skills_excel,skills_sql,skills_soft,family_siblings,family_income_father,family_income_mother,devices_laptop_brand,devices_laptop_year,devices_phone_brand,devices_phone_year
0,1,Claudia Short,20.0,NaN,NaN,NaN,Katieland,NaN,2022-09-03 23:22:44,data science,...,5.0,1.0,"with, onto",4.0,1198.0,1089.0,HP,2021.0,Xiaomi,2021.0
1,2,NaN,20.0,female,90.0,NaN,Dawnburgh,psmith@chen.com,2017-08-29 00:00:00,data science,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,3,Kathryn Moyer,20.0,NaN,90.0,NaN,Lake Stevenmouth,NaN,NaT,data science,...,3.0,1.0,"despite, clearly",1.0,263.0,202.0,HP,2020.0,Samsung,2023.0
3,4,Ruben Wilson,20.0,female,81.0,NaN,Port Pamelafort,NaN,NaT,data science,...,5.0,1.0,"full, series",4.0,508.0,698.0,Lenovo,2020.0,Samsung,2024.0
4,5,Robert Pruitt,20.0,female,NaN,NaN,Kingburgh,NaN,2023-11-25 00:00:00,data science,...,4.0,1.0,"whose, save",3.0,862.0,1058.0,Lenovo,2023.0,Apple,2022.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,996,Gary Sawyer,22.0,female,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992-11-13 00:00:00,data science,...,5.0,1.0,"increase, everything",5.0,1382.0,1335.0,Dell,2017.0,Xiaomi,2019.0
996,997,Scott Clark,20.0,NaN,NaN,+1-205-065-8737,Rogerborough,katiepage@hensley.com,2024-10-05 00:00:00,data science,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
997,998,NaN,20.0,male,NaN,+1-264-577-8585,South Mary,NaN,2021-11-19 02:23:00,data science,...,4.0,4.0,"dinner, while",5.0,1285.0,1005.0,HP,2020.0,Xiaomi,2024.0
998,999,Brittany Barrett,21.0,female,86.0,NaN,Port Rachael,richardgreen@shannon-jenkins.info,2021-07-08 05:51:03,data science,...,2.0,1.0,"present, cup",3.0,541.0,571.0,Dell,2022.0,Samsung,2019.0


#### Step 6: Address parsing

##### Address columnni alohida qism va key-larga ajratish: shahar, tuman, pochta kodi. 
##### Raw address columnni saqlab, yangi columnlar hosil qilish (addr_city, addr_district, addr_postal). 

In [73]:
def parse_address(value):
    if pd.isna(value):
        return pd.Series({'addr_city': np.nan, 'addr_district': np.nan, 'addr_postal': np.nan})
    
    value = str(value).strip()
    
    # BROKEN → all NaN
    if value.startswith('BROKEN'):
        return pd.Series({'addr_city': np.nan, 'addr_district': np.nan, 'addr_postal': np.nan})
    
    # "Apartment 37, South Kevin district, Tashkent, UZ, 100539"
    if value.startswith('Apartment'):
        parts    = [p.strip() for p in value.split(',')]
        district = parts[1].replace(' district', '').strip() if len(parts) > 1 else np.nan
        city     = parts[2].strip()                          if len(parts) > 2 else np.nan
        postal   = parts[4].strip()                          if len(parts) > 4 else np.nan
        return pd.Series({'addr_city': city, 'addr_district': district, 'addr_postal': postal})
    
    # "UZ 100332 Tashkent South Patricia"
    if value.startswith('UZ'):
        match = re.match(r'UZ\s+(\d+)\s+(\w+)\s+(.*)', value)
        if match:
            return pd.Series({
                'addr_city'    : match.group(2),
                'addr_district': match.group(3).strip(),
                'addr_postal'  : match.group(1)
            })
    
    # "Wendyshire 12-kv, dom 1, Tashkent"
    if 'kv' in value:
        parts = [p.strip() for p in value.split(',')]
        return pd.Series({
            'addr_city'    : parts[2].strip() if len(parts) > 2 else np.nan,
            'addr_district': parts[0].split(' ')[0].strip() if parts else np.nan,
            'addr_postal'  : np.nan
        })
    
    # US-style "70576 Butler Harbor Suite 847, Janetberg"
    parts = [p.strip() for p in value.split(',')]
    return pd.Series({
        'addr_city'    : parts[-1].strip() if parts else np.nan,
        'addr_district': np.nan,
        'addr_postal'  : np.nan
    })

# Apply
addr_expanded = df_dirty['address_raw'].apply(parse_address)

# Merge into df_dirty (keep original address_raw)
df_dirty = pd.concat([df_dirty, addr_expanded], axis=1)

# Verify
print(df_dirty[['address_raw', 'addr_city', 'addr_district', 'addr_postal']].head(15))
print('\nFinal shape:           ', df_dirty.shape)
print('addr_city non-null:    ', df_dirty['addr_city'].notna().sum())
print('addr_district non-null:', df_dirty['addr_district'].notna().sum())
print('addr_postal non-null:  ', df_dirty['addr_postal'].notna().sum())

                                          address_raw  addr_city  \
0   Apartment 37, South Kevin district, Tashkent, ...   Tashkent   
1                   UZ 100332 Tashkent South Patricia   Tashkent   
2                   Wendyshire 12-kv, dom 1, Tashkent   Tashkent   
3   Apartment 16, North Tamara district, Tashkent,...   Tashkent   
4                              BROKEN,ADDRESS,DATA,,,        NaN   
5   Apartment 42, North Jamesfurt district, Tashke...   Tashkent   
6              Lake Tracishire 29-kv, dom 8, Tashkent   Tashkent   
7            70576 Butler Harbor Suite 847, Janetberg  Janetberg   
8              Williamsborough 2-kv, dom 18, Tashkent   Tashkent   
9                              BROKEN,ADDRESS,DATA,,,        NaN   
10  Apartment 34, North Samantha district, Tashken...   Tashkent   
11  Apartment 13, Morrismouth district, Tashkent, ...   Tashkent   
12                Frankchester 9-kv, dom 20, Tashkent   Tashkent   
13                  UZ 100394 Tashkent Stevensbo

In [74]:
df_dirty

,student_id,name,age,gender,score,phone,city,email,date_of_join,course,...,family_siblings,family_income_father,family_income_mother,devices_laptop_brand,devices_laptop_year,devices_phone_brand,devices_phone_year,addr_city,addr_district,addr_postal
0,1,Claudia Short,20.0,NaN,NaN,NaN,Katieland,NaN,2022-09-03 23:22:44,data science,...,4.0,1198.0,1089.0,HP,2021.0,Xiaomi,2021.0,Tashkent,South Kevin,100539
1,2,NaN,20.0,female,90.0,NaN,Dawnburgh,psmith@chen.com,2017-08-29 00:00:00,data science,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Tashkent,South Patricia,100332
2,3,Kathryn Moyer,20.0,NaN,90.0,NaN,Lake Stevenmouth,NaN,NaT,data science,...,1.0,263.0,202.0,HP,2020.0,Samsung,2023.0,Tashkent,Wendyshire,NaN
3,4,Ruben Wilson,20.0,female,81.0,NaN,Port Pamelafort,NaN,NaT,data science,...,4.0,508.0,698.0,Lenovo,2020.0,Samsung,2024.0,Tashkent,North Tamara,100097
4,5,Robert Pruitt,20.0,female,NaN,NaN,Kingburgh,NaN,2023-11-25 00:00:00,data science,...,3.0,862.0,1058.0,Lenovo,2023.0,Apple,2022.0,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,996,Gary Sawyer,22.0,female,NaN,NaN,Gabriellefurt,amy97@woodard.net,1992-11-13 00:00:00,data science,...,5.0,1382.0,1335.0,Dell,2017.0,Xiaomi,2019.0,Tashkent,Laurenhaven,100254
996,997,Scott Clark,20.0,NaN,NaN,+1-205-065-8737,Rogerborough,katiepage@hensley.com,2024-10-05 00:00:00,data science,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Tashkent,Patrickmouth,NaN
997,998,NaN,20.0,male,NaN,+1-264-577-8585,South Mary,NaN,2021-11-19 02:23:00,data science,...,5.0,1285.0,1005.0,HP,2020.0,Xiaomi,2024.0,Tashkent,New Brendan,100394
998,999,Brittany Barrett,21.0,female,86.0,NaN,Port Rachael,richardgreen@shannon-jenkins.info,2021-07-08 05:51:03,data science,...,3.0,541.0,571.0,Dell,2022.0,Samsung,2019.0,West Scott,NaN,NaN


#### Step 7: Duplicate va missing data tekshirish 

##### Duplicate rowsni aniqlash va o‘chirish. 

In [75]:
print("DUPLICATES")
print("Total duplicate rows:      ", df_dirty.duplicated().sum())
print("Duplicate by student_id:   ", df_dirty.duplicated(subset='student_id').sum())


print("\n MISSING VALUES ")
missing     = df_dirty.isnull().sum()
missing_pct = (df_dirty.isnull().sum() / len(df_dirty) * 100).round(1)
summary = pd.DataFrame({'missing_count': missing, 'missing_%': missing_pct})
print(summary[summary['missing_count'] > 0].sort_values('missing_%', ascending=False))

DUPLICATES
Total duplicate rows:       0
Duplicate by student_id:    0

 MISSING VALUES 
                      missing_count  missing_%
phone                           739       73.9
addr_postal                     603       60.3
email                           561       56.1
addr_district                   430       43.0
skills_soft                     406       40.6
skills_python                   406       40.6
skills_excel                    406       40.6
family_income_father            406       40.6
family_siblings                 406       40.6
devices_phone_brand             406       40.6
devices_laptop_year             406       40.6
family_income_mother            406       40.6
devices_laptop_brand            406       40.6
devices_phone_year              406       40.6
skills_sql                      406       40.6
name                            335       33.5
gpa                             301       30.1
score                           246       24.6
gender            

In [77]:
dr_dup=df_dirty.drop_duplicates(subset='name')

In [ ]:
before = len(df_dirty)
df_dirty.drop_duplicates(inplace=True)
df_dirty.drop_duplicates(subset='student_id', keep='first', inplace=True)
after = len(df_dirty)

print(f"Rows before: {before}")
print(f"Rows after:  {after}")
print(f"Dropped:     {before - after}")

Rows before: 1000
Rows after:  1000
Dropped:     0


##### Muhim columnlarda missing values mavjudligini aniqlash va qaror qabul qilish (fillna yoki dropna). 

In [ ]:
critical = ['name', 'age', 'gender', 'score', 'gpa']
before = len(df_dirty)
df_dirty.dropna(subset=critical, how='all', inplace=True)
print(f"Dropped {before - len(df_dirty)} rows where all critical fields were missing")

df_dirty['gender'].fillna('Unknown', inplace=True)
df_dirty['remarks'].fillna('No remarks', inplace=True)
df_dirty['attendance'].fillna(df_dirty['attendance'].median(), inplace=True)
df_dirty['score'].fillna(df_dirty['score'].median(), inplace=True)
df_dirty['gpa'].fillna(df_dirty['gpa'].median(), inplace=True)
df_dirty['age'].fillna(df_dirty['age'].median(), inplace=True)


remaining = df_dirty.isnull().sum()
print(remaining[remaining > 0])
print("\nFinal shape:", df_dirty.shape)

Dropped 2 rows where all critical fields were missing
name                    333
age                     106
gender                  217
score                   244
phone                   737
email                   559
date_of_join            167
gpa                     299
remarks                 204
event_time              173
hobbies                 185
skills_python           405
skills_excel            405
skills_sql              405
skills_soft             405
family_siblings         405
family_income_father    405
family_income_mother    405
devices_laptop_brand    405
devices_laptop_year     405
devices_phone_brand     405
devices_phone_year      405
addr_city               211
addr_district           430
addr_postal             603
dtype: int64

Final shape: (998, 32)


C:\Users\user\AppData\Local\Temp\ipykernel_24940\3779496194.py:6: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df_dirty['gender'].fillna('Unknown', inplace=True)
C:\Users\user\AppData\Local\Temp\ipykernel_24940\3779496194.py:7: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment

#### Step 8: Data normalization 

NULL va NaN mavzusida hammasini logical to'g'tilab ketganman, step 3

##### Gender columnini standard formatga keltirish: Male / Female / Unknown. 

In [ ]:
df_dirty['gender'].fillna('Unknown', inplace=True)

C:\Users\user\AppData\Local\Temp\ipykernel_24940\1328258479.py:1: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df_dirty['gender'].fillna('Unknown', inplace=True)


0      Unknown
1       female
2      Unknown
3       female
4       female
        ...   
995     female
996    Unknown
997       male
998     female
999    Unknown
Name: gender, Length: 998, dtype: str

##### Course columnlarini yagona nomga keltirish: Data Science / Python / Other. 

In [ ]:
df_dirty['course'].fillna('Other', inplace=True)

C:\Users\user\AppData\Local\Temp\ipykernel_24940\662003348.py:1: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df_dirty['course'].fillna('Other', inplace=True)


0      data science
1      data science
2      data science
3      data science
4      data science
           ...     
995    data science
996    data science
997    data science
998    data science
999          python
Name: course, Length: 998, dtype: str

##### Status columnlarini standart formatga keltirish (lower-case yoki uniform). 

In [ ]:
df_dirty['status'].fillna('unknown', inplace=True)

C:\Users\user\AppData\Local\Temp\ipykernel_24940\3266027572.py:1: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df_dirty['status'].fillna('unknown', inplace=True)


0        active
1        active
2       pending
3      inactive
4       pending
         ...   
995    inactive
996     pending
997    inactive
998      active
999    inactive
Name: status, Length: 998, dtype: str

#### Step 9: Final type conversion va export 

##### Barcha columnlarning to‘g‘ri type da ekanligini tekshirish: string, int, float, datetime.

In [ ]:
df_dirty.columns

Index(['student_id', 'name', 'age', 'gender', 'score', 'phone', 'city',
       'email', 'date_of_join', 'course', 'attendance', 'status', 'gpa',
       'remarks', 'money_spent', 'event_time', 'address_raw', 'hobbies',
       'skills_python', 'skills_excel', 'skills_sql', 'skills_soft',
       'family_siblings', 'family_income_father', 'family_income_mother',
       'devices_laptop_brand', 'devices_laptop_year', 'devices_phone_brand',
       'devices_phone_year', 'addr_city', 'addr_district', 'addr_postal'],
      dtype='str')

In [ ]:
df_dirty.dtypes

student_id                      int64
name                              str
age                           float64
gender                            str
score                         float64
phone                             str
city                              str
email                             str
date_of_join            datetime64[s]
course                            str
attendance                      int64
status                            str
gpa                           float64
remarks                           str
money_spent                   float64
event_time              datetime64[s]
address_raw                       str
hobbies                           str
skills_python                 float64
skills_excel                  float64
skills_sql                    float64
skills_soft                       str
family_siblings               float64
family_income_father          float64
family_income_mother          float64
devices_laptop_brand              str
devices_lapt

##### Sanalarni yagona formatga keltirish (YYYY-MM-DD HH:MM:SS). 

In [ ]:
df_dirty.info()

<class 'pandas.DataFrame'>
Index: 998 entries, 0 to 999
Data columns (total 32 columns):
 #   Column                Non-Null Count  Dtype        
---  ------                --------------  -----        
 0   student_id            998 non-null    int64        
 1   name                  665 non-null    str          
 2   age                   892 non-null    float64      
 3   gender                781 non-null    str          
 4   score                 754 non-null    float64      
 5   phone                 261 non-null    str          
 6   city                  998 non-null    str          
 7   email                 439 non-null    str          
 8   date_of_join          831 non-null    datetime64[s]
 9   course                998 non-null    str          
 10  attendance            998 non-null    int64        
 11  status                998 non-null    str          
 12  gpa                   699 non-null    float64      
 13  remarks               794 non-null    str          

##### Tozalangan data CSV faylga saqlash (super_dirty_students_cleaned.csv). 

In [ ]:
df_dirty.shape

(998, 32)

In [ ]:
df_dirty.to_csv('super_dirty_students_cleaned2.csv', index=True)
print("\n Saved as super_dirty_students_cleaned.csv")


 Saved as super_dirty_students_cleaned.csv


In [ ]:
#df_gpa=df_dirty['gpa'].groupby(by='city').mean()
avg_gpa_by_city = df_dirty.groupby('city')['gpa'].mean().round(2)
print(avg_gpa_by_city)

city
Aaronberg       2.0
Adrianabury     NaN
Adrienneside    2.0
Aguilarfort     3.7
Albertshire     2.0
               ... 
Youngmouth      NaN
Youngside       NaN
Zacharyside     4.5
Zhangbury       NaN
Zhangstad       3.7
Name: gpa, Length: 967, dtype: float64


#### Step 10: QA checks 

##### Original va cleaned row sonini solishtirish. 

In [ ]:
df_original = pd.read_csv('super_dirty_students.csv')

print(" ROW COUNT ")
print("Original rows: ", len(df_original))
print("Cleaned rows:  ", len(df_dirty))
print("Dropped rows:  ", len(df_original) - len(df_dirty))

print("\n DUPLICATES")
print("Duplicate rows:      ", df_dirty.duplicated().sum())
print("Duplicate student_id:", df_dirty.duplicated(subset='student_id').sum())
print(" No duplicates!" if df_dirty.duplicated().sum() == 0 else "Duplicates found")

 ROW COUNT 
Original rows:  1000
Cleaned rows:   998
Dropped rows:   2

 DUPLICATES
Duplicate rows:       0
Duplicate student_id: 0
 No duplicates!


##### Missing email va phone sonini tekshirish. 

In [ ]:
print("Missing email: ", df_dirty['email'].isna().sum(), 
      f"({df_dirty['email'].isna().mean()*100:.1f}%)")
print("Missing phone: ", df_dirty['phone'].isna().sum(), 
      f"({df_dirty['phone'].isna().mean()*100:.1f}%)")
print("Missing both:  ", df_dirty[df_dirty['email'].isna() & df_dirty['phone'].isna()].shape[0])

Missing email:  559 (56.0%)
Missing phone:  737 (73.8%)
Missing both:   415


##### Numeric columnlar (GPA, attendance, score) qiymatlari to‘g‘ri diapazonda ekanligini tekshirish.

hammasi joyida :)

##### Duplicate rowlar yo‘qligini tasdiqlash. 

hammasi joyida :)